Add new block commands

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Advances: #51559
Adds block selection related commands: block transforms and block duplicate, copy, remove, insert after/before etc.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Mainly to improve the usefulness of the command center, but also because this improve the distraction free mode experience by offering immediate access to basic functions without turn off the mode.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Register the commands using mostly copied functionality from the block switcher and block actions components.

## Testing Instructions

1. Select some block
2. Press cmd + k 
3. Type "transform to" or "copy"
4. See command
5. Execute command
6. It should do what it says

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/107534/b02d9053-41db-4816-8178-3299fbb3d61b

## To do

- [x] Does this make things slower? I notice some lagginess but it may be just me.
- [x] Cleanup some of the copy pasta (there's mentions of patterns there).
- [x] Is the block list the best place to register these commands?
- [x] Should these commands use a context?

draganescu committed:
inital attempt to bring block commands and transforms

draganescu committed:
don't register commands if no blocks are selected

youknowriad commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -40,13 +40,15 @@ import {
 const pendingBlockVisibilityUpdatesPerRegistry = new WeakMap();
 
 function Root( { className, ...settings } ) {
+	useBlockCommands();
```
It's not clear to me whether this should be called here. What happens when there are multiple block editors in the page. (Previews are a block editor for instance)

draganescu replied:
Yes. I have the same doubt. Where would you say a better place to register these would be?

youknowriad replied:
Probably top level in edit-post and edit-site

draganescu replied:
Done, let me know it it's all right now as placement for init of commands registering.

draganescu committed:
move to top level editor components

draganescu committed:
cleanup copy pasta from the block editor components

youknowriad commented on packages/block-editor/src/components/index.js:
```diff
@@ -165,6 +165,7 @@ export { default as __experimentalInspectorPopoverHeader } from './inspector-pop
 
 export { default as BlockEditorProvider } from './provider';
 export { default as useSetting } from './use-setting';
+export { useBlockCommands } from './use-block-commands';
```
Do we need the `Block` in the name of this hook. It's already part of block editor so maybe it can be just `useCommands` which would stand for (use all block editor commands)

draganescu replied:
I think it's fine to be granular? Is there any downside?

youknowriad commented on packages/block-editor/src/components/use-block-commands/index.js:
```diff
@@ -0,0 +1,292 @@
+		name: 'core/block-editor/blockActions',
+		hook: useActionsCommands,
+	} );
+};
```
Both of these command loaders use a lot of selectors... I do wonder about a potential performance impact on the command palette with long posts. Maybe it's fine though. One follow-up here could be to add a performance metric about the command palette.

draganescu committed:
fix rebase artefact

draganescu commented:
@youknowriad do you think this can land as is? Should I try and improve anything in particular? I see the tests (including perf.) don't complain.

youknowriad commented:
Sure but let's get more testing first.

jameskoster commented:
The only commands I see are the transforms. No Copy, Move, or any of the other examples from the video in the OP. Is that correct?

I think it might be good to try displaying the contextual actions as suggestions immediately after opening the command palette, rather than having to search for them:

<img width="909" alt="Screenshot 2023-07-17 at 14 24 00" src="https://github.com/WordPress/gutenberg/assets/846565/9f804351-4e2e-46f1-85fa-eeb327001932">

Obviously we'll need to work out a mechanism to scale this as we add more actions, likely drilldowns:

<img width="1115" alt="Screenshot 2023-07-17 at 14 31 58" src="https://github.com/WordPress/gutenberg/assets/846565/74c1e7e1-2c45-4cfa-9e0b-c7e1a5eac155">

But that's a detail to figure out separately.

draganescu commented:
@jameskoster I think you should see transforms and some of the basic block actions like move, insert before etc. I tested to see if the PR is having some issue and apparently it's all right couldn't repro. Maybe its your build?

In terms of UI - when do we know to suggest what?

jameskoster commented:
Still only seeing the transform commands 🤔 Tried both post and site editors. 

> In terms of UI - when do we know to suggest what?

Worth a sanity check from @WordPress/gutenberg-design but I think it might be good to surface actions that apply to the selected block when the command palette is invoked. If you can detail exactly which commands this PR will add, we can work out priorities etc.

draganescu commented:
@jameskoster this PR adds:

- all transforms _to_ the block has defined (e.g. to cover, to gallery, to columns, to file, to group, to media and text, for an image block)
- these block actions: paste styles, copy, ungroup, group, moveTo, insertAfter, insertBefore, remove, duplicate)

Showing all of them by default may be overwhelming.

jameskoster commented:
> Showing all of them by default may be overwhelming.

Agreed. Since the aim of this PR is to add so many commands, let's not surface any suggestions yet. We can explore that in a follow-up with a thought-out design which considers how to scale the display of so many commands, if necessary.

richtabor commented:
Works well! Some quick bits, that could either be here or in a follow-up: 

1. I don't think we need to append "block" to the end of every command. We don't have them on the controls within the toolbar options either. 

2. The UI uses "Add before" and "Add after" instead of "Insert":
<img width="257" alt="CleanShot 2023-07-24 at 11 52 39" src="https://github.com/WordPress/gutenberg/assets/1813435/b50faf85-1dd9-43f0-a5a7-e6fe509e6dce">
<img width="476" alt="CleanShot 2023-07-24 at 11 51 05" src="https://github.com/WordPress/gutenberg/assets/1813435/14da39fe-245b-4a44-98f8-e174f46dc1c9">

3. I'm ok omitting "Move to", it's not connecting well triggered from the palette. 

5. Do commands require an icon? If not, I think we should remove the fallback "copy" icon being applied to every one. 

6. cc @WordPress/gutenberg-design — should block titles be capitalized here? 
<img width="477" alt="CleanShot 2023-07-24 at 11 54 43" src="https://github.com/WordPress/gutenberg/assets/1813435/4024ba26-af40-4b0d-b3c9-0596e971629b">

draganescu committed:
update icons to match the action, remove 'block' suffix for command names, use add instead of insert

draganescu commented:
I've rebased and updated the PR with the comments from @richtabor .

The icons part was a bit tricky as they are required by I made do with the icons we have. I don't think this to be a blocker as any visual tweaks can come in a follow up.

I suggest we merge this to see if in the real world it causes any issues.

Also I left the move to command on because it is extremely useful in distraction free mode, with the caveat that it looks bad :)

draganescu committed:
remove conflict marker

scruffian commented:
Transform is working fine, and copying appears to work, but when I paste, the clipboard just contains the same thing as before.

mtias commented on packages/block-editor/src/components/use-block-commands/index.js:
```diff
@@ -0,0 +1,317 @@
+
+	/**
+	 * The `isTemplate` check is a stopgap solution here.
+	 * Ideally, the Transforms API should handle this
```
Seems it'd be better to capture this comment in a separate issue instead

draganescu replied:
I think that is copy pasta from the block tools code :)

draganescu commented:
@scruffian does it work using the tools in the block menu? Maybe there is a focus issue with copying?

richtabor commented:
> The icons part was a bit tricky as they are required by I made do with the icons we have. I don't think this to be a blocker as any visual tweaks can come in a follow up.

I don't think they should required: https://github.com/WordPress/gutenberg/issues/53193

draganescu commented:
@richtabor is that blocking this? I think the icons I use here are good enough.

richtabor commented:
Not blocking no. I'm taking another look right now.

richtabor commented:
I tried to get them to pull up but I wasn't able to (in post or site editor). Another rebase perhaps? 😬

richtabor reviewed (approved):
Works well; could use a bit of refining in a follow up: 
- #53538
- #53539

draganescu committed:
packages update

scruffian committed:
fix copy command

scruffian commented on packages/block-editor/src/components/use-block-commands/index.js:
```diff
@@ -0,0 +1,324 @@
+		replaceBlocks( clientIds, innerBlocks );
+	};
+	const onCopy = () => {
+		Clipboard.copy( serialize( blocks ) );
```
I added this line so the blocks are actually copied.

scruffian commented on packages/compose/package.json:
```diff
@@ -38,7 +38,7 @@
 		"@wordpress/priority-queue": "file:../priority-queue",
 		"change-case": "^4.1.2",
-		"clipboard": "^2.0.8",
+		"clipboard": "^2.0.11",
```
We need to update the version of clipboard we are using so that the `.copy` function is available

scruffian commented on packages/block-editor/package.json:
```diff
@@ -66,6 +67,7 @@
 		"@wordpress/wordcount": "file:../wordcount",
 		"change-case": "^4.1.2",
 		"classnames": "^2.3.1",
+		"clipboard": "2.0.11",
```
This is a new dependency.

youknowriad replied:
In other places, we just use `event.clipboardData.setData`

draganescu committed:
Revert "fix copy command"

This reverts commit e6d70b232ac4fda45133cf50ae8ee6d0996cef44.

draganescu committed:
remove copy and paste styles actions

draganescu commented:
I have reverted [6662c28](https://github.com/WordPress/gutenberg/pull/52509/commits/6662c28b7d0dd313bce60576d5eb5844363e5cc9) and removed copy and paste styles.

The idea is to land the two hooks here with actions and transforms and see if there are any issues (perf) with them. For the copy and paste actions I'll follow up in another PR so we can have the time to spend on it without holding the rest back.

richtabor commented:
Very cool. This would be to feature in an upcoming Gutenberg release; perhaps after https://github.com/WordPress/gutenberg/issues/53538 and https://github.com/WordPress/gutenberg/issues/53539.

youknowriad commented on packages/block-editor/src/components/use-block-commands/index.js:
```diff
@@ -0,0 +1,284 @@
+				Array.isArray( clientIds ) ? clientIds[ 0 ] : clientIds
+			);
+			return {
+				possibleBlockTransformations: getBlockTransformItems(
```
In some contexts, this is triggering failures in the command loader. I'm guessing when the "blocks" array is empty.

cc @draganescu 

Also @artpi for more details.

artpi replied:
I'm still tracking down the exact reproduction path, but here is a stack trace I am getting when I run command loader quite often (for a different env that command pallette):

![image](https://github.com/WordPress/gutenberg/assets/3775068/af07eb71-b095-4a0b-a7a9-7126d40717c0)

draganescu replied:
I am not sure what to do to trigger a visible error. Ultimately empty `blocks` is guarded in `getPossibleBlockTransformations` by:

```
if ( ! blocks.length ) {
		return [];
	}
```

artpi replied:
Yes, I see that in the code and yet I get the above stack trace. My suspicion is that we're getting `[null, null ]` or something similar

artpi replied:
I finally have a reproduciton path. It took me quite a while:

https://github.com/WordPress/gutenberg/issues/59289
